orchestrator/src/coordinator.rs

```diff
@@ -1,30 +1,128 @@
+use snafu::prelude::*;
 use std::{
     collections::{HashMap, VecDeque},
     convert::{TryFrom, TryInto},
     path::{Path, PathBuf},
     process::Stdio,
-    str::from_utf8,
+    str::from_utf8, sync::atomic::{AtomicU64, Ordering},
 };
-
-use snafu::prelude::*;
 use tokio::{
-    io::{AsyncReadExt, AsyncWriteExt, BufReader},
     process::{Child, ChildStdin, ChildStdout, Command},
     select,
-    sync::mpsc,
-    task::JoinSet,
+    sync::{mpsc, oneshot},
+    task::{JoinSet, JoinHandle},
 };
+use tokio_util::io::SyncIoBridge;
 
 use crate::{
     message::{CommandId, CoordinatorMessage, Job, JobReport, WorkerMessage},
     sandbox::{CompileRequest, CompileResponse},
 };
 
-pub type Container = (
-    Child,
-    mpsc::Sender<PlaygroundMessage>,
-    mpsc::Receiver<ContainerMessage>,
-);
+#[derive(Debug)]
+enum FanoutCommand {
+    Listen(u64, mpsc::Sender<ContainerMessage>),
+}
+
+#[derive(Debug)]
+pub struct Container {
+    child: Child,
+    tx: mpsc::Sender<PlaygroundMessage>,
+    command_tx: mpsc::Sender<FanoutCommand>,
+    id: AtomicU64,
+}
+
+impl Container {
+    async fn fanout(mut command_rx: mpsc::Receiver<FanoutCommand>, mut rx: mpsc::Receiver<ContainerMessage>) {
+        let mut waiting = HashMap::new();
+
+        loop {
+            select! {
+                command = command_rx.recv() => {
+                    let command = command.expect("Handle this");
+                    match command {
+                        FanoutCommand::Listen(id, waiter) => {
+                            waiting.insert(id, waiter);
+                            // TODO: ensure not replacing
+                        }
+                    }
+                },
+
+                msg = rx.recv() => {
+                    let msg = msg.expect("Handle this");
+                    let id = 0; // TODO: some uniform way of getting the ID for _any_ message
```

Comment:
Perhaps the messages between the coordinator and the worker should be of the form 

`(Id, ActualMessage)`, so that every message has an ID baked in.